### PR TITLE
drivers: gpio_nrfx: Introduce minimal implementation

### PR DESCRIPTION
**Pull request by MarekPieta**
Change introduces a new configuration option that can be used to disable GPIO interrupts support to reduce memory footprint.

### CODE REVIEW

**Review by gmarull — CHANGES_REQUESTED** [review]
sounds like this should be a driver class option, ie, disable interrupt support

**Comment by MarekPieta** [comment]
The minimal configuration may be useful e.g. if GPIO is used by MCUboot bootloader to enter serial recovery mode. In that case GPIO interrupts are not needed. Enabling the new Kconfig option allows to reduce FLASH usage by ~700 bytes.

**Comment by MarekPieta** [comment]
Hi @gmarull,
I agree that eventually disabling interrupt support should be a driver class option.

Still, the changes does not break the current API (according to documentation: `gpio_pin_interrupt_configure` is allowed to return `-ENOTSUP` if a requested interrupt configuration is not supported). Also, updating GPIO driver API greatly increases scope of the change. Could we keep the change as is for now and update it after a proper driver class option is introduced? We need this change to fit MCUboot bootloader in one of our configurations.

**Comment by gmarull** [comment]
> Hi @gmarull, I agree that eventually disabling interrupt support should be a driver class option.
> 
> Still, the changes does not break the current API (according to documentation: `gpio_pin_interrupt_configure` is allowed to return `-ENOTSUP` if a requested interrupt configuration is not supported). Also, updating GPIO driver API greatly increases scope of the change. Could we keep the change as is for now and update it after a proper driver class option is introduced? We need this change to fit MCUboot bootloader in one of our configurations.

The fact that it requires a bit more work doesn't mean it can't be done. I don't think this is a big job, probably just some ifdeffery, and update of APIs to return `-ENOSYS` if not available.

**Review by mkapala-nordic — APPROVED** [review]

**Comment by MarekPieta** [comment]
There seem to be already quite a few GPIO drivers that introduced their own Kconfig option to enable/disable interrupt support (e.g. `CONFIG_GPIO_NCT38XX_INTERRUPT`, `CONFIG_GPIO_SX1509B_INTERRUPT`, `CONFIG_GPIO_PCA95XX_INTERRUPT`, etc.). I aligned my PR with the Kconfig option naming convention to improve consistency. My change does not introduce inconsistency comparing with the existing code, it also shouldn't break anything. Could we review and merge it as is and then discuss the generic GPIO Kconfig under a separate PR (then we could align nrfx driver too)?

Adding a generic Kconfig option for GPIO interrupt support control would require changes in all of the GPIO driver implementations (the generic Kconfig option would be used to reduce functionality and memory footprints for the GPIO drivers, so drivers must take it into account in the implementation). We would also need approvals from the codeowners. There is also a risk that bigger change could accidentally break something (I am not able to test the change with most of the available GPIO drivers). I guess that the generic change may also trigger discussions, which could delay introducing change to the nrfx driver (the fix introduced here is urgent for us).

**Review by anangl — DISMISSED** [review]

**Review by mkapala-nordic — APPROVED** [review]

File: drivers/gpio/gpio_nrfx.c
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_nrfx.c:406]: Wont this lead to compilation warnings about `gpio_nrfx_manage_callback()` being unused if `CONFIG_GPIO_NRFX_INTERRUPT=n`?

File: drivers/gpio/gpio_nrfx.c
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_nrfx.c:247]: Wouldn't an `#ifdef CONFIG_GPIO_NRFX_INTERRUPT ... #else ... #endif` be better here?

File: drivers/gpio/gpio_nrfx.c
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_nrfx.c:391]: Same here?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_nrfx.c
Comment (MarekPieta) [reply drivers/gpio/gpio_nrfx.c:406]: We use `IS_ENABLED`, not `ifdefs` here so that the unused compilation warnings should not happen here (I did not see them locally too)

File: drivers/gpio/gpio_nrfx.c
Comment (MarekPieta) [reply drivers/gpio/gpio_nrfx.c:247]: I prefer to use `IS_ENABLED` approach to keep the compiler checks here (code is opt out from the resulting code anyway so we still save memory)

File: drivers/gpio/gpio_nrfx.c
Comment (MarekPieta) [reply drivers/gpio/gpio_nrfx.c:391]: Please check the answer above

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by gmarull — DISMISSED** [review]
I'm fine leaving this option (will try to submit a generalized Kconfig option when I have time), but please, refactor this PR following https://github.com/zephyrproject-rtos/zephyr/pull/62394

**Comment by MarekPieta** [comment]
Hi @gmarull,
I applied the changes. Please rereview

**Review by anangl — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
